daemon/metrics.go

```diff
@@ -134,3 +137,117 @@ func init() {
+//     regenerations may be rolled up into one by a trigger.Trigger.
+//   - The revision only ever increments and we can never see an older revision
+//     once we have successfully waited on it.
+// The buffered pendingMetricRevisions channel is used as the input to this
+// goroutine and holds an entry for every PolicyAdd call that resulted in a new
+// revision. It is expected to remain fairly empty and only fill up when
+// multiple revision changes queue up during a slower regeneration. Closing the
+// channel causes the goroutine to exit.
+func startImplementationDelayWatcher() (pendingMetricRevisions chan policyRevChange) {
+	// pendingMetricRevisions is a buffered channel of revisions for which we
+	// want the implementation delay.
+	// It is returned after the goroutine is started below and used by
+	// recordPolicyDelayMetric.
+	pendingMetricRevisions = make(chan policyRevChange, 16384)
+
+	// Start a goroutine to wait on revision changes sequentially, then report
+	// the PolicyImplementationDelay metric for each entry seen in
+	// pendingMetricRevisions
+	go func() {
+		log.Debug("startImplementationDelayWatcher goroutine started")
+		defer log.Debug("startImplementationDelayWatcher goroutine exited")
+
+		var (
+			// starts is a map of revision -> metadata for revisions in
+			// pendingMetricRevisions that were in the future when returning from
+			// WaitForEndpointsAtPolicyRevision. Once we get up to that revision the
+			// entry is handled and deleted
+			starts = make(map[uint64]policyRevChange)
+
+			// currentRevision is what we are currently operating on. It is only
+			// incremented once we succcessfully wait on that revision number and
+			// record metrics for it.
+			currentRevision = uint64(0)
+		)
+
+		for {
+			// backoff if there are no endpoints. This is required to avoid
+			// spin-looping and incrementing the revision
+			// (WaitForEndpointsAtPolicyRevision returns when there are no
+			// endpoints).
+			if len(endpointmanager.GetEndpoints()) == 0 {
+				time.Sleep(10 * time.Second)
+				continue
+			}
+
+			// Wait for currentRevision and, once it returns, for each entry in pendingMetricRevisions:
+			//  - If pendingMetricRevisions is closed, exit the goroutine
+			//  - If the desired revision is in the past, record the metric immediately
+			//  - If the revision is in the future, store it for when that revision is seen
+			// Break out when pendingMetricRevisions is empty
+			endpointmanager.WaitForEndpointsAtPolicyRev(context.Background(), currentRevision)
+			log.WithField(logfields.PolicyRevision, currentRevision).Debug("startImplementationDelayWatcher finished waiting for a revision")
+			now := time.Now()
+		pendingRevisions:
+			for {
+				select {
+				case start, stillOpen := <-pendingMetricRevisions:
+					switch {
+					case !stillOpen:
+						return
+					case start.rev <= currentRevision:
+						duration := now.Sub(start.timestamp)
```

Comment:
Does this need to use `spanstat` to not run into the negative metric problem?

Reply:
Good point. I added a guard but I think histograms won't panic (it was a counter the last time and those are supposed to only go up).